CommonMark/ExtensibleHtmlPrinter.cs

```diff
@@ -0,0 +1,742 @@
+                    case BlockTag.FencedCode:
+                        PrintCode(writer, block);
+                        break;
+
+                    case BlockTag.HtmlBlock:
+                        PrintHtmlBlock(writer, block);
+                        break;
+
+                    case BlockTag.HorizontalRuler:
+                        PrintHorizontalRuler(writer, block);
+                        break;
+
+                    case BlockTag.ReferenceDefinition:
+                        break;
+
+                    default:
+                        throw new CommonMarkException("Block type " + block.Tag + " is not supported.", block);
+                }
+
+                if (_visitChildren)
+                {
+                    _stack.Push(new HtmlPrinter.BlockStackEntry(_stackLiteral, block.NextSibling, _tight));
+
+                    _tight = _stackTight;
+                    block = block.FirstChild;
+                }
+                else if (block.NextSibling != null)
+                {
+                    block = block.NextSibling;
+                }
+                else
+                {
+                    block = null;
+                }
+
+                while (block == null && _stack.Count > 0)
+                {
+                    var entry = _stack.Pop();
+
+                    WriteLineConstant(entry.Literal);
+                    _tight = entry.IsTight;
+                    block = entry.Target;
+                }
+            }
+        }
+
+        protected virtual void PrintHorizontalRuler(TextWriter writer, Block block)
+        {
+            if (_trackPosition)
+            {
+                WriteConstant("<hr");
+                PrintPosition(block);
+                writer.WriteLine();
+            }
+            else
+            {
+                WriteLineConstant("<hr />");
+            }
+        }
+
+        protected virtual void PrintHtmlBlock(TextWriter writer, Block block)
+        {
+            // cannot output source position for HTML blocks
+            block.StringContent.WriteTo(writer);
+        }
+
+        protected virtual void PrintCode(TextWriter writer, Block block)
+        {
+            int x;
+            EnsureLine();
+            WriteConstant("<pre><code");
+            if (_trackPosition) PrintPosition(block);
+
+            var info = block.FencedCodeData == null ? null : block.FencedCodeData.Info;
+            if (info != null && info.Length > 0)
+            {
+                x = info.IndexOf(' ');
+                if (x == -1)
+                    x = info.Length;
+
+                WriteConstant(" class=\"language-");
+                EscapeHtml(new StringPart(info, 0, x));
+                writer.Write('\"');
+            }
+            Write('>');
+            EscapeHtml(block.StringContent);
+            WriteLineConstant("</code></pre>");
+        }
+
+        protected virtual void PrintHeader(TextWriter writer, Block block, CommonMarkSettings settings)
```

Comment:
`writer` and `settings` should be properties, not passed to every method.
Also I think that the `writer` might not be exposed at all for now - just the most minimal set of methods (such as `Write(string)`, `WriteLine(string)`).